### PR TITLE
Setting default values for 'health_settings' attribute

### PR DESCRIPTION
**Pull request by ziemowit141**
https://clkentik.atlassian.net/browse/KNTK-309

I've set default values for `health_settings`

Tested in 3 cases:
1) empty `health_settings` passed
2) no `health_settings` passed
3) `health_settings` filled up

Tested with actual Kentik API

### CODE REVIEW

**Review by mmac-m3a — APPROVED** [review]
LGTM

**Comment by mmac-m3a** [comment]
Waiting for Daniel to approve it too.

File: synthetics/test_conversions.go [hunk 1]
```diff
@@ -467,11 +467,23 @@ func resourceDataToTestHealthSettings(data interface{}) (*synthetics.V202101beta
 	if err != nil {
 		return nil, fmt.Errorf("resourceDataToTestHealthSettings: %v", err)
 	}
+
+	obj := synthetics.NewV202101beta1HealthSettings()
+
 	if m == nil {
-		return nil, nil
+		obj.SetLatencyCritical(0)
```
Comment (Fenthick) [inline synthetics/test_conversions.go:474]: Please add a comment that Kentik API requires health_settings object

File: synthetics/test_conversions.go [hunk 2]
```diff
@@ -467,11 +467,23 @@ func resourceDataToTestHealthSettings(data interface{}) (*synthetics.V202101beta
...
+		obj.SetLatencyWarning(0)
+		obj.SetPacketLossCritical(0)
+		obj.SetPacketLossWarning(0)
+		obj.SetJitterCritical(0)
+		obj.SetJitterWarning(0)
+		obj.SetHttpLatencyCritical(0)
+		obj.SetHttpLatencyWarning(0)
+		obj.SetHttpValidCodes([]int64{})
+		obj.SetDnsValidCodes([]int64{})
+		return obj, nil
```
Comment (Fenthick) [inline synthetics/test_conversions.go:484]: When user specifies no health_settings, test resource can be created. However, the subsequent apply fails with following error:

```shell
➜  kentik-synthetics_test git:(pstroz/KNTK-309) ✗ terraform apply
kentik-synthetics_test.example-hostname-test: Refreshing state... [id=5296]

Note: Objects have changed outside of Terraform

Terraform detected the following changes made outside of Terraform since the last "terraform apply":

  # kentik-synthetics_test.example-hostname-test has been changed
  ~ resource "kentik-synthetics_test" "example-hostname-test" {
        id              = "5296"
        name            = "example-hostname-test"
        # (6 unchanged attributes hidden)

      ~ settings {
            # (13 unchanged attributes hidden)



          ~ monitoring_settings {
              + notification_channels   = []
                # (4 unchanged attributes hidden)
            }


            # (4 unchanged blocks hidden)
        }
    }

Unless you have made equivalent changes to your configuration, or ignored the relevant attributes using ignore_changes, the following plan may include actions to undo or respond to these changes.

─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # kentik-synthetics_test.example-hostname-test will be updated in-place
  ~ resource "kentik-synthetics_test" "example-hostname-test" {
        id              = "5296"
        name            = "example-hostname-test"
        # (6 unchanged attributes hidden)

      ~ settings {
            # (13 unchanged attributes hidden)

          - health_settings {
              - dns_valid_codes       = [] -> null
              - http_latency_critical = 0 -> null
              - http_latency_warning  = 0 -> null
              - http_valid_codes      = [] -> null
              - jitter_critical       = 0 -> null
              - jitter_warning        = 0 -> null
              - latency_critical      = 0 -> null
              - latency_warning       = 0 -> null
              - packet_loss_critical  = 0 -> null
              - packet_loss_warning   = 0 -> null
            }




            # (4 unchanged blocks hidden)
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.

Changes to Outputs:
  ~ hostname-test = {
      ~ settings        = [
          ~ {
              ~ health_settings     = [
                  - {
                      - dns_valid_codes       = []
                      - http_latency_critical = 0
                      - http_latency_warning  = 0
                      - http_valid_codes      = []
                      - jitter_critical       = 0
                      - jitter_warning        = 0
                      - latency_critical      = 0
                      - latency_warning       = 0
                      - packet_loss_critical  = 0
                      - packet_loss_warning   = 0
                    },
                ]
              ~ monitoring_settings = [
                  ~ {
                      ~ notification_channels   = null -> []
                        # (4 unchanged elements hidden)
                    },
                ]
                # (23 unchanged elements hidden)
            },
        ]
        # (9 unchanged elements hidden)
    }

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

kentik-synthetics_test.example-hostname-test: Modifying... [id=5296]
╷
│ Error: failed to patch test
│ 
│   with kentik-synthetics_test.example-hostname-test,
│   on resource.tf line 1, in resource "kentik-synthetics_test" "example-hostname-test":
│    1: resource "kentik-synthetics_test" "example-hostname-test" {
│ 
│ error: 400 Bad Request, status: 400 Bad Request, body: {{"code":3,"message":"Invalid update mask (no such field 'test.settings.health_settings.http_valid_codes') (errxid c4qtaesp64h0c6rbdz6g)","details":[]}}
╵

```

The provider sends a patch request with all health_settings fields in field mask (`commonPatchTestFields`), however empty list of HttpValidCodes is discarded by JSON.Marshal _omitempty_ option. API Server fails with such payload.

> The "omitempty" option specifies that the field should be omitted from the encoding if the field has an empty value, defined as false, 0, a nil pointer, a nil interface value, and any empty array, slice, map, or string. 
https://pkg.go.dev/encoding/json#Marshal

**Review by Fenthick — CHANGES_REQUESTED** [review]

File: synthetics/test_conversions.go (hunk 2 above)
Comment (ziemowit141) [reply synthetics/test_conversions.go:484]: @Fenthick First of all I'm surprised by your thoughtfulness. How did you know to check such a case? Maybe we should create some test suit, that will check all corner cases, so we can simply run it while doing any changes on terraform configuration?

Comment (mmac-m3a) [reply synthetics/test_conversions.go:484]: A more comprehensive test suite, testing full life cycle (create, modify, delete), would be nice. However, it would requite deeper emulation of Kentik API, allowing to emulate sequence of responses. Note that automated development testing against the actual API is undesirable.

Is there any way to insert default valuea so that the Terraform engine believes that it came from the user? If not, let's not sweat it. We'd though need to include those defaults in examples, so that users have a place to copy from. ... and again, Kentik API will get better over time allowing to omit attributes that can be safely defaulted (or are really not relevant in the context of the very request - e.g. HTTPValidCodes for a normal ping/trace test).

We must support passing empty lists for HTTP and DNS result codes.

Comment (ziemowit141) [reply synthetics/test_conversions.go:475]: It turned out that the problem is twofold.

**Problem 1.** Kentik API requires `health_settings`. When we pass empty `health_settings` in a POST request, the server assigns 0 values to `health_settings` attributes in just created test resource.

**Payload when creating a test resource:**

> {"@level":"info","@message":"2021/09/08 11:58:53 \nPOST /synthetics/v202101beta1/tests HTTP/1.1\r\nHost: synthetics.api.kentik.com\r\nUser-Agent: OpenAPI-Generator/1.0.0/go\r\nContent-Length: 721\r\nAccept: application/json\r\nContent-Type: application/json\r\nX-Ch-Auth-Api-Token: f60b0f96c43ccd7454308384e64d9023\r\nX-Ch-Auth-Email: piotr.stroz@codilime.com\r\nAccept-Encoding: gzip\r\n\r\n{\"test\":{\"deviceId\":\"0\",\"name\":\"agents-filtered-by-country-test\",\"settings\":{\"agentIds\":[\"968\",\"970\",\"926\",\"900\",\"967\",\"1208\",\"2764\"],\"count\":0,\"expiry\":0,\"family\":\"IP_FAMILY_V6\",\"healthSettings\":{},\"hostname\":{\"target\":\"www.example.com\"},\"limit\":0,\"monitoringSettings\":{\"activationGracePeriod\":\"2\",\"activationTimeUnit\":\"m\",\"activationTimeWindow\":\"5\",\"activationTimes\":\"3\",\"notificationChannels\":[]},\"period\":0,\"ping\":{\"count\":5,\"expiry\":3000,\"period\":60},\"port\":443,\"protocol\":\"icmp\",\"reciprocal\":false,\"rollupLevel\":1,\"servers\":[],\"tasks\":[\"ping\",\"traceroute\"],\"trace\":{\"count\":3,\"expiry\":22500,\"limit\":30,\"period\":60,\"port\":33434,\"protocol\":\"udp\"},\"useLocalIp\":false},\"status\":\"TEST_STATUS_ACTIVE\",\"type\":\"hostname\"}}","@timestamp":"2021-09-08T11:58:53.674985+02:00"}

**Payload when reading just created test resource:**

> {"@level":"info","@message":"2021/09/08 12:01:03 \nHTTP/2.0 200 OK\r\nContent-Length: 1142\r\nContent-Type: application/json\r\nDate: Wed, 08 Sep 2021 10:01:03 GMT\r\nGrpc-Accept-Encoding: identity\r\nGrpc-Encoding: identity\r\nGrpc-Message: OK\r\nGrpc-Status: 0\r\nReferrer-Policy: strict-origin-when-cross-origin\r\nServer: envoy\r\nStrict-Transport-Security: max-age=31536000; includeSubDomains; preload;\r\nX-B3-Traceid: aee0ca69cebb7beb\r\nX-Content-Type-Options: nosniff\r\nX-Envoy-Upstream-Service-Time: 33\r\nX-Frame-Options: DENY\r\nX-Ratelimit-Details: eyJkZXNjcmlwdG9yX3N0YXR1c2VzIjpbeyJjb2RlIjoxLCJjdXJyZW50X2xpbWl0Ijp7InJlcXVlc3RzX3Blcl91bml0IjoxODAwLCJ1bml0IjoyfSwibGltaXRfcmVtYWluaW5nIjoxNzk5LCJkdXJhdGlvbl91bnRpbF9yZXNldCI6eyJzZWNvbmRzIjo1N319XX0\r\nX-Request-Id: 74074ae1-1138-9748-8d2b-185b0c0235ac\r\nX-Xss-Protection: 1; mode=block\r\n\r\n{\"test\":{\"id\":\"5383\",\"name\":\"agents-filtered-by-country-test\",\"type\":\"hostname\",\"deviceId\":\"75702\",\"status\":\"TEST_STATUS_ACTIVE\",\"settings\":{\"hostname\":{\"target\":\"www.example.com\"},\"agentIds\":[\"968\",\"970\",\"926\",\"900\",\"967\",\"1208\",\"2764\"],\"period\":0,\"count\":0,\"expiry\":0,\"limit\":0,\"tasks\":[\"ping\",\"traceroute\"],\"healthSettings\":{\"latencyCritical\":0,\"latencyWarning\":0,\"packetLossCritical\":0,\"packetLossWarning\":0,\"jitterCritical\":0,\"jitterWarning\":0,\"httpLatencyCritical\":0,\"httpLatencyWarning\":0,\"httpValidCodes\":[],\"dnsValidCodes\":[]},\"monitoringSettings\":{\"activationGracePeriod\":\"2\",\"activationTimeUnit\":\"m\",\"activationTimeWindow\":\"5\",\"activationTimes\":\"3\",\"notificationChannels\":[]},\"ping\":{\"period\":60,\"count\":5,\"expiry\":3000,\"delay\":0},\"trace\":{\"period\":60,\"count\":3,\"protocol\":\"udp\",\"port\":33434,\"expiry\":22500,\"limit\":30,\"delay\":0},\"port\":443,\"protocol\":\"icmp\",\"family\":\"IP_FAMILY_V6\",\"servers\":[],\"targetType\":\"\",\"targetValue\":\"\",\"useLocalIp\":false,\"reciprocal\":false,\"rollupLevel\":1},\"createdBy\":{\"id\":\"194579\",\"email\":\"piotr.stroz@codilime.com\",\"fullName\":\"\"},\"cdate\":\"2021-09-08T09:58:59.950Z\",\"edate\":\"2021-09-08T09:58:59.950Z\"}}","@timestamp":"2021-09-08T12:01:03.500477+02:00"}

The important part is that in the first payload, `health_settings` are empty, in the second they are set to 0 values.

**Problem 2.** PATCH method field mask is not updated with optional arguments.

This is the actual root cause of the error @Fenthick found. `health_settings` were considered a required field in a field mask. Request failed when the server received a payload with `health_settings` in the field mask but without corresponding values. My most recent commit fixed that issue.

Conclusions:
After fixing the fields mask, the feature, that allows `health_settings` to be optional in Terraform works, although end-user will get notified about: `Note: Objects have changed outside of Terraform` because of the mismatch between terraform configuration and values stored in Kentik servers. So the functionality is there, the presentation could be better.

I see two possible approaches here:
1. While reading test resources from the Kentik API, Interpret `health_settings` set to 0 as nil, and hopefully, make terraform happy.
2. Leave this task in current condition, after Kentik API gets updated with optional parameters, the problem will solve itself as both terraform configuration and current state on the server will be matching.

@mmac-m3a Please comment on that.

File: synthetics/resource_test_impl.go [hunk 3]
```diff
@@ -196,6 +186,50 @@ func optionalPatchTestFields(test *synthetics.V202101beta1Test) []string {
 		fields = append(fields, "test.settings.url.target")
 	}
 
+	if test.Settings.HasServers() && len(*test.Settings.Servers) != 0 {
```
Comment (Fenthick) [inline synthetics/resource_test_impl.go:189]:
```suggestion
	if test.Settings.HasServers() {
```

I think that user should have a possibility to update server list from non-zero elements to zero elements, i.e. delete elements.

File: synthetics/test_conversions.go (hunk 2 above)
Comment (Fenthick) [reply synthetics/test_conversions.go:475]: I vote for option number 2, which is currently implemented in this PR. I added one small comment - apart from that I think that it is good solution.

**Review by Fenthick — COMMENTED** [review]
One small comment

File: synthetics/resource_test_impl.go (hunk 3 above)
Comment (Fenthick) [reply synthetics/resource_test_impl.go]: Done

File: synthetics/test_conversions.go [hunk 4]
```diff
@@ -143,11 +143,41 @@ func testURLToMapSlice(obj *synthetics.V202101beta1UrlTest) []map[string]interfa
...
+		obj.JitterCritical == nil &&
+		obj.JitterWarning == nil &&
+		obj.HttpLatencyCritical == nil &&
+		obj.HttpLatencyWarning == nil &&
+		obj.HttpValidCodes == nil &&
+		obj.DnsValidCodes == nil {
+		return nil
+	}
+
+	// Kentik API sets 0 values for nil fields
```
Comment (Fenthick) [inline synthetics/test_conversions.go:165]:
```suggestion
	// Kentik API sets 0 values for omitted optional fields.
```
I would avoid using _nil_ while describing JSON payloads.

File: synthetics/test_conversions.go [hunk 5]
```diff
@@ -143,11 +143,41 @@ func testURLToMapSlice(obj *synthetics.V202101beta1UrlTest) []map[string]interfa
...
+	// Necessary conversion to nil, so Terraform configuration
+	// matches with actual state on the server
+	if *obj.LatencyCritical == 0 &&
+		*obj.LatencyWarning == 0 &&
```
Comment (Fenthick) [inline synthetics/test_conversions.go:169]:
```suggestion
		obj.GetLatencyWarning() == 0 &&
```
Please use getters here to avoid the possibility of nil pointer dereference.

**Review by Fenthick — CHANGES_REQUESTED** [review]
Almost there ;)

File: synthetics/test_conversions.go (hunk 4 above)
Comment (ziemowit141) [reply synthetics/test_conversions.go]: Addressed: 0e8ee0d8730ca72fcd78be1b5f7aa7a7f5c0e604

File: synthetics/test_conversions.go (hunk 5 above)
Comment (ziemowit141) [reply synthetics/test_conversions.go]: Addressed: 0e8ee0d8730ca72fcd78be1b5f7aa7a7f5c0e604

**Review by Fenthick — APPROVED** [review]

**Comment by Fenthick** [comment]
@mmac-m3a Please review and merge
